pact-core/Pact/Core/IR/Eval/CoreBuiltin.hs

```diff
@@ -148,20 +153,49 @@ rawAdd info b cont handler _env = \case
     chargeGasArgs info (GConcat (ListConcat (GasListLength (V.length l + V.length r))))
     returnCEKValue cont handler (VList (l <> r))
   args -> argsError info b args
+  where
+  decimalAdd i i' = do
+    chargeGasArgs info (GIntegerOpCost PrimOpAdd (decimalMantissa i) (decimalMantissa i'))
+    returnCEKValue cont handler (VLiteral (LDecimal (i + i')))
 
 rawSub :: (CEKEval step b i m, MonadEval b i m) => NativeFunction step b i m
 rawSub info b cont handler _env = \case
-  [VLiteral (LInteger i), VLiteral (LInteger i')] -> returnCEKValue cont handler (VLiteral (LInteger (i - i')))
-  [VLiteral (LDecimal i), VLiteral (LDecimal i')] -> returnCEKValue cont handler (VLiteral (LDecimal (i - i')))
+  [VLiteral (LInteger i), VLiteral (LInteger i')] -> do
+    chargeGasArgs info (GIntegerOpCost PrimOpAdd i i')
+    returnCEKValue cont handler (VLiteral (LInteger (i - i')))
+  -- Overloaded decimal cases
+  [VLiteral (LDecimal i), VLiteral (LDecimal i')] ->
+    decimalSub i i'
+  [VLiteral (LInteger i), VLiteral (LDecimal i')] ->
+    decimalSub (Decimal 0 i) i'
+  [VLiteral (LDecimal i), VLiteral (LInteger i')] ->
+    decimalSub i (Decimal 0 i')
   args -> argsError info b args
+  where
+  decimalSub i i' = do
+    chargeGasArgs info (GIntegerOpCost PrimOpAdd (decimalMantissa i) (decimalMantissa i'))
```

Comment:
Should this be `PrimOpAdd`? Addition and subtraction are indeed very similar, so if that's intended, I'd probably rename `PrimOpAdd` → `PrimOpAddSub`.

Reply:
Great suggestion!! Will do